### ISSUE TITLE
Synchronize GitHub Source Code Mappings to Java Build Tool Model

### ISSUE THREAD

**Issue by gilday** [getsentry/sentry-android-gradle-plugin#546]
### Problem Statement

I would like Sentry tooling to synchronize the source code mappings in my Sentry GitHub integration with my Java project structure.

Consider a multi-project Gradle project with subprojects `foo`, `bar`, and `baz`. In this case, I would like an automated way to create source code mappings for directories `foo/src/main/java`, `bar/src/main/java`, and `baz/src/main/java`.

### Solution Brainstorm

Java build tool (e.g. Maven and Gradle) are aware of projects' directory structures and the difference between production and test code. A Sentry Maven / Gradle plugin could query the project's structure and synchronize the GitHub source code mappings.

### Product Area

Issues

**Comment by romtsn** [comment]
I've moved this to the SAGP repo where it should belong I believe. We'll have to investigate quarkus and see what it does (I've got a feeling that something they are doing is off, but maybe it's also us).

**Comment by adinauer** [comment]
Hello @gilday have you had a chance to test the fix @lbloder suggested?

**Comment by Glyphack** [comment]
Hi @adinauer I had the same issue and this comment fixed it:
https://github.com/getsentry/sentry-android-gradle-plugin/issues/546#issuecomment-1713950054

My issue was not exactly this but similarly a tool was generating code at build time and I had to add this dependency. I think we can add this to troubleshooting or setup gradle plugin docs. I came here from this page:
https://docs.sentry.io/platforms/java/guides/logback/gradle/

**Comment by adinauer** [comment]
Thanks for the feedback @Glyphack we'll improve troubleshooting docs.

**Comment by Glyphack** [comment]
I can give it a try.

**Comment by gilday** [comment]
Can confirm that workaround is effective.

**Comment by gilday** [comment]
It's my understanding that the Gradle plugin uploads sources to provide source context. With this working, I can see my sources alongside stack traces, but there are still no links to GitHub.

The feature I described in the description of this issue pertains to GitHub [code mappings](https://docs.sentry.io/platforms/java/source-context/#setting-up-code-mappings) which feels different than only seeing sources.

Is it accurate to say that the Gradle plugin provides source context but does not help with configuring GitHub code mappings?

**Comment by romtsn** [comment]
@gilday so, the source context is a prerequisite for displaying github links afaik, otherwise even if you set up correct code mappings, they just won't be shown in the UI without the source context. So you need both in this case, and the gradle plugin helps with the former. 

Have you tried setting up code mappings in your project settings as describe in the guide you linked?

**Comment by gilday** [comment]
I have tried setting up those code mappings. They're not working, but I now see that code mapping requires a Business payment tier but I have a Team payment tier at the moment.

I'm not likely to upgrade for this feature alone, because manually configuring code mappings for my multi-module Gradle project is too tedious. I have dozens of projects in my multi-module build, and each one needs its own code mapping. The mappings I set-up in August are already woefully out of date with the current state of the repository.

This brings me to the original new capability request in the description of this issue: I'd like for the Sentry Gradle plugin to manage my code mappings for this project. Without that, the code mapping feature gives me more toil than value.

**Comment by adinauer** [comment]
Not sure how this works exactly in the Sentry backend but we could try to make it smarter by recognizing popular build tools (Gradle, Maven) and auto mapping modules after scanning the repo.

**Comment by adinauer** [comment]
Ah thanks for the info. Good thing we already have a gradle and maven plugin, maybe they just need to be told the repo URL and offer some task that can set up mappings.

**Comment by wzieba** [comment]
hi there 👋 I believe this issue should not be closed: it started with 

> tooling to synchronize the source code mappings in my Sentry GitHub integration with my Java project structure

but was closed with a commit mentioning documentation about setting dependency with code generation tasks.

To circle back to the original request from @gilday, I'd like to second to 

> I have dozens of projects in my multi-module build, and each one needs its own code mapping. The mappings I set-up in August are already woefully out of date with the current state of the repository.

this is very on point 🎯 . I guess it's unlike that a team of multimodule Android project would invest into maintaining dozens of Code Mappings manually; therefore they lose plenty of great features you prepared (Stack Trace Links, Source Contexts etc.).

Would you consider:
- introducing a feature to Sentry Gradle Plugin which would provide Code Mappings 
- or at lest API for this, so we could prepare a script ourselves
- or any other solution to this?

**Comment by adinauer** [comment]
Sorry this was closed via the PR that added docs for a fix mentioned in this issue. Reopening.

**Comment by adinauer** [comment]
@wzieba afaict there's currently no API to configure code mappings. So achieving this seems a little more comlex than assumed.

**Comment by wzieba** [comment]
Thank you @adinauer for opening this again! 

> there's currently no API to configure code mappings

It's unfortunate :(. It'd be really great if there was one - ideally, mapping would be associated with the specific build via some kind of id (the same as Source Context or Proguard mappings are) as mapping can change from build to build.

I hope this feature will be prioritized some day.

**Comment by adinauer** [comment]
Yeah we'll likely have to make this more generic and rework integrations as well. It's now back in our backlog but I can't say when we'll get to it.

**Comment by adinauer** [comment]
@wzieba you could take a look at the Sentry UI in browser dev tools and inspect API requests performed when adding code mappings, then replicate this in a script.

**Comment by wzieba** [comment]
Good idea @adinauer ! I've come up with something like this:

<details><summary>custom task</summary>

```kotlin
tasks.register("createSentryMappings") {
    doLast {
        fun makeRequest(dryRun: Boolean, sourceRoot: String, stackRoot: String) {
            println("About to send mapping\nsourceRoot: $sourceRoot\nstackRoot: $stackRoot\n---")
            if (dryRun) return

            val body = HttpRequest.BodyPublishers.ofString(
                """
                    {
                      "defaultBranch": "main",
                      "integrationId": "<id>",
                      "projectId": "<id>",
                      "repositoryId": "<id>",
                      "sourceRoot": "$sourceRoot",
                      "stackRoot": "$stackRoot"
                    }
                """.trimIndent()
            )
            val request = HttpRequest
                .newBuilder(URI("https://us.sentry.io/api/0/organizations/<org>/code-mappings/"))
                .setHeader(
                    "Authorization",
                    "Bearer <token>"
                )
                .setHeader("Content-Type", "application/json")
                .POST(body)
                .build()

            val response =
                HttpClient.newHttpClient().send(request, HttpResponse.BodyHandlers.ofString())

            logger.info(response.body())
        }
        rootProject.allprojects
            .filter {
                it.name !in listOf(
                    rootProject.name,
                    <projects_to_exclude>
                )
            }
            .forEach {
                makeRequest(
                    false,
                    sourceRoot = "modules/${it.project.parent!!.name}/${it.project.name}/src/main/java/au/com/shiftyjelly/pocketcasts/${it.project.name}",
                    stackRoot = "au/com/shiftyjelly/pocketcasts/${it.project.name}",
                )
            }
    }
}

```

</details>

It's tailored for my use case with hard-coded source and stack roots, but I guess it can be a starting point for someone else looking for a way to automate sending mappings.

**Comment by adinauer** [comment]
Thanks @wzieba that's great 🚀
